src/Tasks/Exec.cs

```diff
@@ -609,20 +612,53 @@ protected internal override void AddCommandLineCommands(CommandLineBuilderExtens
                     }
                     commandLine.AppendSwitch("/C"); // run then terminate
 
-                    // If for some crazy reason the path has a & character and a space in it
-                    // then get the short path of the temp path, which should not have spaces in it
-                    // and then escape the &
-                    if (batchFileForCommandLine.Contains("&") && !batchFileForCommandLine.Contains("^&"))
+                    if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
                     {
-                        batchFileForCommandLine = NativeMethodsShared.GetShortFilePath(batchFileForCommandLine);
-                        batchFileForCommandLine = batchFileForCommandLine.Replace("&", "^&");
+                        StringBuilder fileName = StringBuilderCache.Acquire(batchFileForCommandLine.Length);
+
+                        // Escape special characters that need to be escaped.
+                        for (int i = 0; i < batchFileForCommandLine.Length; i++)
+                        {
+                            char c = batchFileForCommandLine[i];
+
+                            if (ShouldEscapeCharacter(c) && (i == 0 || batchFileForCommandLine[i - 1] != '^'))
+                            {
+                                fileName.Append('^');
+                            }
+                            fileName.Append(c);
+                        }
+                        batchFileForCommandLine = StringBuilderCache.GetStringAndRelease(fileName);
+                    }
+                    else
+                    {
+                        // If for some crazy reason the path has a & character and a space in it
+                        // then get the short path of the temp path, which should not have spaces in it
+                        // and then escape the &
+                        if (batchFileForCommandLine.Contains("&") && !batchFileForCommandLine.Contains("^&"))
+                        {
+                            batchFileForCommandLine = NativeMethodsShared.GetShortFilePath(batchFileForCommandLine);
+                            batchFileForCommandLine = batchFileForCommandLine.Replace("&", "^&");
+                        }
                     }
                 }
 
                 commandLine.AppendFileNameIfNotNull(batchFileForCommandLine);
             }
         }
 
+        private bool ShouldEscapeCharacter(char c)
+        {
+            for (int i = 0; i < _charactersToEscape.Length; i++)
+            {
+                if (c == _charactersToEscape[i])
+                {
+                    return true;
+                }
+            }
+
+            return false;
```

Comment:
nit: Wondering how `IndexOf` would compare to the manual loop in terms of perf for such a small array.

```suggestion
            return Array.IndexOf(_charactersToEscape, c) != -1;
```

Reply:
Did some rough testing here. On small cases (~200 iterations) the array performed marginally faster (about ~500 ticks). Array.IndexOf performs better under more iterations. But this was only when run ~10k+ times, which paths aren't likely to get to.

The more realistic case seems to be 150~250 characters, depending on temp and how long someone's username is. Will keep it as a for loop through the array.

Reply:
Note the "average" case looks something like this `C:\\Users\\username\\AppData\\Local\\Temp\\sh4yq5zn.hpc\\Temporary4476f4a8231944f0959781eebe164cdb\\hello()w]o(rld)\\tmpb547f707b5e34e4e9f9b5d49ac496be4.exec.cmd`